### PR TITLE
feat(parquet): reduce and limit buffering when writing parquet files

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/7783
Closes https://github.com/influxdata/influxdb_iox/pull/7866
Closes https://github.com/influxdata/influxdb_iox/pull/7855
Closes https://github.com/influxdata/influxdb_iox/issues/7860

# TLDR
I believe this PR will
1.  Reduce the memory required for compaction significantly (it should now remain much cluster to the limits configured) 2. It will *NOT* change the size of the generated parquet files

# Rationale:
As explained on https://github.com/influxdata/influxdb_iox/issues/7783#issuecomment-1557789373 the parquet writer in `parquet 40.0.0` buffers up to 1M rows prior to encoding a row group, regardless of the amount of memory that consumes. For some types of data, this amount of memory can be 10s of GB which causes the compactor to be OOM killed.

# Changes
1. Patch and use a (pre-release) version of `parquet` and arrow that contains https://github.com/apache/arrow-rs/pull/4280 and uses *significantly* less memory. See https://github.com/alamb/arrow-rs/pull/37 for full contents
1. Track the memory usage used in the parquet writer in the DataFusion `MemoryPool`. This  means that if the total memory used by DataFusion and the write exceeds the limit set by the pool the parquet write may fail.

# Effect on Memory Usage

## On this branch
While writing out the parquet files from the scenario on https://github.com/influxdata/influxdb_iox/issues/7783 with this change, the memory profile is consistent and stays under 2GB. Here is a screen shot from `Instruments`:

Screenshot 2023-05-30 at 1 46 36 PM

## main (causes k8s OOM kills)
On `main`, the memory usage peaks at 20+GB (picture from https://github.com/influxdata/influxdb_iox/issues/7783#issuecomment-1557713255):
Screenshot 2023-05-22 at 2 35 11 PM

# Effect on Parquet File Size
None

### CODE REVIEW

File: parquet_file/src/writer.rs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+        // drop the writer and verify the pool is freed
+        assert_ne!(pool.reserved(), 0);
+        assert_ne!(pool.reserved(), writer.inner.in_progress_size());
+        std::mem::drop(writer);
+        assert_eq!(pool.reserved(), 0);
+    }
+
+    #[tokio::test]
+    async fn test_pool_memory_pressure() {
```
Comment (alamb) [inline parquet_file/src/writer.rs:163]: I expect that this test passes, however, it fails because the reported memory allocation size is always 0.

Comment (alamb) [reply parquet_file/src/writer.rs:163]: Will work out with @tustvold  upstream: https://github.com/apache/arrow-rs/pull/4280/files#r1207342693

**Review by Joe-Blount — APPROVED** [review]
LGTM, but I'm just along for the ride on this one...

File: Cargo.toml [hunk 2]
```diff
@@ -128,6 +128,22 @@ tonic-build = { version = "0.9.2" }
 tonic-health = { version = "0.9.2" }
 tonic-reflection = { version = "0.9.2" }
 
+[patch.crates-io]
```
Comment (crepererum) [inline Cargo.toml:131]: is this planned to land w/ a patch (that's fine, just checking)?

**Review by crepererum — APPROVED** [review]

File: parquet_file/src/writer.rs [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+        let reservation = consumer.register(&pool);
+
+        Ok(Self { inner, reservation })
+    }
+
+    /// Push a `RecordBatch` into the underlying writer, updating the
+    /// tracked allocation
+    pub fn write(&mut self, batch: RecordBatch) -> Result<()> {
+        // writer encodes the batch into its internal buffers
+        let result = self.inner.write(&batch);
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet_file/src/writer.rs:64]: ArrowWriter will splice the batch if it is too large

File: Cargo.toml (hunk 2 above)
Comment (alamb) [reply Cargo.toml:131]: Yes, that is the plan -- to use a patched version until arrow 41.0.0 is released (eta around June 5 or June 6 2023)

File: parquet_file/src/writer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (alamb) [reply parquet_file/src/writer.rs:64]: > This assumes that the batches that DF produces have a reasonable size, right?

In general, I agree the memory limiting behavior assumes the batches are reasonable sizes

**Comment by alamb** [comment]
Update here is that the compactor is still OOM'ing but we think there may be another source of memory usage (the scratchpad) that may be contributing

I claim avoiding intermittent memory spikes for tables with high cardinality tags (which is what this PR does) is still a help

**Pull request merged**